### PR TITLE
ByteBufUtil writeUtf8 and writeAscii helper methods

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Utility methods in ByteBufUtil to writeUtf8 and writeAscii expect a buffer to already be allocated. If the user does not have a buffer allocated they have to know details of the encoding in order to know the size of the buffer to allocate.

Modifications:
- Add writeUtf8 and writeAscii which take a ByteBufAllocator and allocate a ByteBuf of the correct size for the user

Result:
ByteBufUtil methods which are easier to use if the user doesn't already have a ByteBuf.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer @nmittler - FYI

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 1]
```diff
@@ -355,21 +355,29 @@ private static int lastIndexOf(ByteBuf buffer, int fromIndex, int toIndex, byte
...
      * Encode a {@link CharSequence} in <a href="http://en.wikipedia.org/wiki/UTF-8">UTF-8</a> and write
+     * it to a {@link ByteBuf} allocated with {@code alloc}.
+     * @param alloc The allocator used to allocate a new {@link ByteBuf}.
+     * @param seq The characters to write into a buffer.
+     * @return The {@link ByteBuf} which contains the <a href="http://en.wikipedia.org/wiki/UTF-8">UTF-8</a> encoded
+     * result.
+     */
+    public static ByteBuf writeUtf8(ByteBufAllocator alloc, CharSequence seq) {
+        // UTF-8 uses max. 3 bytes per char, so calculate the worst case.
+        ByteBuf buf = alloc.buffer(seq.length() * 3);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: i think we need to ensure releasing the buffer on an error

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 2]
```diff
@@ -437,18 +445,27 @@ private static int writeUtf8(AbstractByteBuf buffer, CharSequence seq, int len)
...
+     * Encode a {@link CharSequence} in <a href="http://en.wikipedia.org/wiki/ASCII">ASCII</a> and write
+     * it to a {@link ByteBuf} allocated with {@code alloc}.
+     * @param alloc The allocator used to allocate a new {@link ByteBuf}.
+     * @param seq The characters to write into a buffer.
+     * @return The {@link ByteBuf} which contains the <a href="http://en.wikipedia.org/wiki/ASCII">ASCII</a> encoded
+     * result.
+     */
+    public static ByteBuf writeAscii(ByteBufAllocator alloc, CharSequence seq) {
+        // ASCII uses 1 byte per char
+        ByteBuf buf = alloc.buffer(seq.length());
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: See above

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 1 above)
Comment (nmittler) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: +1

Comment (nmittler) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: Also, is the constant not defined somewhere already?

Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: > i think we need to ensure releasing the buffer on an error

writeUtf8 no longer throws to be consistent with `String.getBytes()`.

> Also, is the constant not defined somewhere already?

I don't believe so. I can define one locally in this file.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 2 above)
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: `writeAscii` should not throw to be consistent with `getBytes()`. If it throws isn't that an error?

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: true...

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (770e74a01c15b7ed757a605e6f846805ab6bf6b3) and 4.1 (94f27be59b637c74763cc3b9fd27cf628b59f7bd)

**Comment by normanmaurer** [comment]
@Scottmitch thanks buddy!

**Comment by Scottmitch** [comment]
@nmittler @normanmaurer - Thanks for review / cherry-pick.

**Pull request closed without merging**